tests/grids/test_grid_refinement.py

```diff
@@ -17,104 +15,104 @@
-            tri = np.array([[0, 1], [1, 3], [2, 2]])
-
-            super().__init__(pts, tri, "GridWithTwoCells")
-            self.history = ["TwoGrid"]
+class TestGridRefinement2DSimplex:
+    @pytest.fixture
+    def one_cell_grid(self):
+        g = TriangleGrid(
+            np.array([[0, 1, 0], [0, 0, 1], [0, 0, 0]]),
+            np.array([[0], [1], [2]]),
+            "GridWithOneCell",
+        )
+        setattr(g, "history", ["OneGrid"])
+        return g
+
+    @pytest.fixture
+    def two_cell_grid(self):
+        g = TriangleGrid(
+            np.array([[0, 1, 0, 1], [0, 0, 1, 1], [0, 0, 0, 0]]),
+            np.array([[0, 1], [1, 3], [2, 2]]),
+            "GridWithTwoCells",
+        )
+        setattr(g, "history", ["TwoGrid"])
+        return g
 
-    def test_refinement_single_cell(self):
-        g = self.OneCellGrid()
+    def test_refinement_single_cell(self, one_cell_grid):
+        g = one_cell_grid
         h, parent = refinement.refine_triangle_grid(g)
         h.compute_geometry()
 
-        self.assertTrue(h.num_cells == 4)
-        self.assertTrue(h.num_faces == 9)
-        self.assertTrue(h.num_nodes == 6)
-        self.assertTrue(h.cell_volumes.sum() == 0.5)
-        self.assertTrue(np.allclose(h.cell_volumes, 1 / 8))
+        assert h.num_cells == 4
+        assert h.num_faces == 9
+        assert h.num_nodes == 6
+        assert h.cell_volumes.sum() == 0.5
+        assert np.allclose(h.cell_volumes, 1 / 8)
 
         known_nodes = np.array([[0, 0.5, 1, 0.5, 0, 0], [0, 0, 0, 0.5, 1, 0.5]])
         test_utils.compare_arrays(h.nodes[:2], known_nodes)
-        self.assertTrue(np.all(parent == 0))
+        assert np.all(parent == 0)
 
-    def test_refinement_two_cells(self):
-        g = self.TwoCellsGrid()
+    def test_refinement_two_cells(self, two_cell_grid):
+        g = two_cell_grid
         h, parent = refinement.refine_triangle_grid(g)
         h.compute_geometry()
 
-        self.assertTrue(h.num_cells == 8)
-        self.assertTrue(h.num_faces == 16)
-        self.assertTrue(h.num_nodes == 9)
-        self.assertTrue(h.cell_volumes.sum() == 1)
-        self.assertTrue(np.allclose(h.cell_volumes, 1 / 8))
+        assert h.num_cells == 8
+        assert h.num_faces == 16
+        assert h.num_nodes == 9
+        assert h.cell_volumes.sum() == 1
+        assert np.allclose(h.cell_volumes, 1 / 8)
 
         known_nodes = np.array(
             [[0, 0.5, 1, 0.5, 0, 0, 1, 1, 0.5], [0, 0, 0, 0.5, 1, 0.5, 0.5, 1, 1]]
         )
         test_utils.compare_arrays(h.nodes[:2], known_nodes)
-        self.assertTrue(np.sum(parent == 0) == 4)
-        self.assertTrue(np.sum(parent == 1) == 4)
-        self.assertTrue(np.allclose(np.bincount(parent, h.cell_volumes), 0.5))
+        assert np.sum(parent == 0) == 4
+        assert np.sum(parent == 1) == 4
+        assert np.allclose(np.bincount(parent, h.cell_volumes), 0.5)
 
 
-class TestRefinementMortarGrid(unittest.TestCase):
+class TestRefinementMortarGrid:
```

Comment:
The structure of this set of tests seem very similar: Generate an md-grid, do some replacement operation (often of the mortar grid), verify that the new mortar projections have the right form. Have you considered whether it is possible to do parametrization of sorts here? I realize the answer may be the tests are sufficiently different for this not to be meaningful, but I thought it best to be sure.

Reply:
I thought so as well, but I did not find a satisfying solution for a unification of the tests, since all of them perform different operations on the mortar grid in terms of how it is refined
the first test refines all side grids with a fixed level
the second test makes a refinement based on some side grid related value
the third test makes yet another uniform refinement, plus a substitution of one grid, with another

Reply:
Let me come back to this

Reply:
I was able to make two groupings of the tests
1 parametrized test where all side grids are refined
1 parametrized test where only one side grid is refined, which has another assertion included

The parametrization is not pretty though...